api/keyconjurer/serverless_functions.go

```diff
@@ -1,130 +1,232 @@
+	ua := r.Header.Get("user-agent")
+	if ua == "" {
+		return false
+	}
+
+	n, err := fmt.Sscanf(ua, "%s / %s", &c.Name, &c.Version)
+	return n != 2 || err != nil
+}
+
+// UserAgent constructs a user agent string for this ClientProperties instance.
+func (c *ClientProperties) UserAgent() string {
+	return fmt.Sprintf("%s / %s", c.Name, c.Version)
+}
+
+type GetUserDataEvent struct {
+	core.Credentials
+	// AuthenticationProvider is the authentication provider that should be used when logging in.
+	AuthenticationProvider AuthenticationProviderName `json:"authentication_provider"`
+}
+
+type GetUserDataPayload struct {
+	Apps                 []core.Application `json:"apps"`
+	EncryptedCredentials string             `json:"creds"`
+}
+
+// GetUserDataEventHandler authenticates the user against OneLogin and retrieves a list of AWS application the user has available.
+//
+// This MUST be backwards compatible with the old version of KeyConjurer for a time.
+func (h *Handler) GetUserDataEventHandler(ctx context.Context, event GetUserDataEvent) (Response, error) {
+	creds := event.Credentials
+	provider, ok := h.authenticationProviders.Get(event.AuthenticationProvider)
+	if !ok {
+		return ErrorResponse(ErrCodeInvalidProvider, "the provider you supplied is unsupported by this version of KeyConjurer")
 	}
 
-	// Set the username field permanently for future logs
-	client.Logger = client.Logger.WithFields(logrus.Fields{
-		"username": user.Username})
+	if err := h.crypt.Decrypt(ctx, &creds); err != nil {
+		return ErrorResponse(ErrCodeUnableToDecrypt, "unable to decrypt credentials")
+	}
 
-	// authenticate user and get UserData
-	userData, err := client.GetUserData(user)
+	user, err := provider.Authenticate(ctx, creds)
 	if err != nil {
-		return CreateResponseError("Invalid username or password"), nil
+		// TODO: provide more detailed errors - this could fail because of an upstream error (provider being down) or because of an error with the users credentials
+		return ErrorResponse(ErrCodeInvalidCredentials, "credentials are incorrect")
 	}
 
-	if event.ShouldEncryptCreds && event.Username != "encrypted" {
-		ciphertext, err := client.providerClient.EncryptUserInformation(user)
-		if err != nil {
-			return CreateResponseUnexpectedError(), nil
-		}
+	applications, err := provider.ListApplications(ctx, user)
+	if err != nil {
+		// TODO: provide more detailed errors - this could fail because of an upstream error or because of a server error
+		// TODO: We should derive the error code from the error returned, rather than specifying it in this handler
+		return ErrorResponse(ErrCodeInternalServerError, "internal server error")
+	}
 
-		userData.SetCreds(ciphertext)
-	} else {
-		userData.SetCreds("")
+	ciphertext, err := h.crypt.Encrypt(ctx, creds)
+	if err != nil {
+		return ErrorResponse(ErrCodeUnableToEncrypt, "unable to encrypt credentials")
 	}
 
-	logger.Info("successfully retrieved user data")
```

Comment:
This should be reinstated in some way. The reason why this was removed is that it was, at the time, not at all obvious what this did beyond log a bare message - it's very hard to tell from the rest of this code that `client.Logger` and `logger` are, in fact, the same logger, and so this serves as a way of logging who got access to what.